app/plugin_signature.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package app
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/mattermost/mattermost-server/mlog"
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/openpgp"
+	"golang.org/x/crypto/openpgp/armor"
+)
+
+var armoredSignatureHeaders = [...]string{
+	"BEGIN PGP MESSAGE",
+	"BEGIN PGP PUBLIC KEY BLOCK",
+	"BEGIN PGP SIGNATURE",
+	"BEGIN PGP SIGNED MESSAGE",
+	"BEGIN PGP ARMORED FILE", /* gnupg extension */
+	"BEGIN PGP PRIVATE KEY BLOCK",
+	"BEGIN PGP SECRET KEY BLOCK", /* only used by pgp2 */
+}
+
+// VerifyPlugin checks that the given signature corresponds to the given plugin and matches a trusted certificate.
+func (a *App) VerifyPlugin(plugin, signature io.Reader) *model.AppError {
+	if verifySignature(bytes.NewReader(mattermostPublicKey), plugin, signature) == nil {
+		return nil
+	}
+	publicKeys, appErr := a.GetPluginPublicKeys()
+	if appErr != nil {
+		return appErr
+	}
+	for _, pk := range publicKeys {
+		pkBytes, appErr := a.GetPublicKey(pk)
+		if appErr != nil {
+			mlog.Error("Unable to get public key for ", mlog.String("filename", pk))
+			continue
+		}
+		publicKey := bytes.NewReader(pkBytes)
+		if verifySignature(publicKey, plugin, signature) == nil {
+			return nil
+		}
+	}
+	return model.NewAppError("VerifyPlugin", "api.plugin.verify_plugin.app_error", nil, "", http.StatusInternalServerError)
```

Comment:
should this be an `InternalServerError` or `Forbidden` or `Unauthorized`, thoughts?